File: patterns/01-design-patterns/03-behavioral/01-strategy/go/strategy.go
package strategy

import "sort"

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type Strategy interface {
	Select([]map[string]any) (string, error)
}
type metricStrategy struct{ field string }

func (s metricStrategy) Select(candidates []map[string]any) (string, error) {
	healthy := make([]map[string]any, 0)
	for _, value := range candidates {
		if value["healthy"] == true {
			healthy = append(healthy, value)
		}
	}
	if len(healthy) == 0 {
		return "", &PatternError{code: "no_healthy_candidate"}
	}
	sort.Slice(healthy, func(i, j int) bool {
		a, b := healthy[i][s.field].(float64), healthy[j][s.field].(float64)
		if a == b {
			return healthy[i]["name"].(string) < healthy[j]["name"].(string)
		}
		return a < b
	})
	return healthy[0]["name"].(string), nil
}
func Evaluate(input map[string]any) (any, error) {
	raw, _ := input["selections"].([]any)
	selected := make([]string, 0, len(raw))
	for _, item := range raw {
		value := item.(map[string]any)
		var current Strategy
		switch value["strategy"] {
		case "cost":
			current = metricStrategy{"cost"}
		case "latency":
			current = metricStrategy{"latency"}
		default:
			return nil, &PatternError{code: "unsupported_strategy"}
		}
		rawCandidates, _ := value["candidates"].([]any)
		candidates := make([]map[string]any, 0, len(rawCandidates))
		for _, candidate := range rawCandidates {
			candidates = append(candidates, candidate.(map[string]any))
		}
		result, err := current.Select(candidates)
		if err != nil {
			return nil, err
		}
		selected = append(selected, result)
	}
	return map[string]any{"selected": selected}, nil
}

File: patterns/01-design-patterns/03-behavioral/01-strategy/python/strategy.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class MetricStrategy:
    field=""
    def select(self,candidates):
        healthy=[value for value in candidates if value.get("healthy",False)]
        if not healthy:raise PatternError("no_healthy_candidate")
        return min(healthy,key=lambda value:(value[self.field],value["name"]))["name"]
class CostStrategy(MetricStrategy):field="cost"
class LatencyStrategy(MetricStrategy):field="latency"
def evaluate(input_data):
    strategies={"cost":CostStrategy(),"latency":LatencyStrategy()};selected=[]
    for value in input_data.get("selections",[]):
        strategy=strategies.get(value.get("strategy"))
        if strategy is None:raise PatternError("unsupported_strategy")
        selected.append(strategy.select(value.get("candidates",[])))
    return{"selected":selected}
